### PR TITLE
fix(cache): add configurable refresh buffer to StorageCredentialCache

### LINKED ISSUE #2046: `StorageCredentialCache` can provide unusable credentials
### Describe the bug

The code design of `StorageCredentialCache` ensures that credentials are cached for some defined amount of time. Returned cached credentials are guaranteed to be valid when retrieved from the cache, but there's no guarantee that the returned credentials will be valid for the next N minutes. It can happen that the vended credentials returned to clients are not or not long enough valid.

### To Reproduce

_No response_

### Actual Behavior

_No response_

### Expected Behavior

_No response_

### Additional context

_No response_

### System information

_No response_

### PR DESCRIPTION
**Pull request by visit2rahul**
Closes #2046

The cache currently evicts credentials after half their remaining
lifetime. That works, but there's no way to tune it -- if your storage
provider returns short-lived tokens, you can end up serving credentials
that are nearly expired by the time the client uses them.

This adds STORAGE_CREDENTIAL_REFRESH_BUFFER_SECONDS (default 0). When
set to a positive value, credentials are evicted from the cache once
their remaining validity drops below that threshold. With the default of
0, behavior is unchanged -- the cache still uses the half-lifetime
approach.

Changes:
- New FeatureConfiguration constant with a clear description and a safe
  default
- StorageCredentialCacheEntry carries the buffer alongside
  maxCacheDurationMs so the Caffeine expiry function can use both
- Two new unit tests: one verifying near-expiry credentials are evicted
  with a 300s buffer, one confirming the zero-buffer path preserves the
  existing behavior

### CODE REVIEW

File: polaris-core/src/test/java/org/apache/polaris/core/storage/cache/StorageCredentialCacheTest.java [hunk 1]
```diff
@@ -442,6 +443,74 @@ private static List<StorageAccessConfig> getFakeAccessConfigs(int number, boolea
...
+        buildKey(
+            "bufferZeroConfig",
+            true,
+            Set.of("s3://bucket/path"),
+            Set.of(),
+            Optional.empty(),
+            () -> config);
+
+    storageCredentialCache.getOrLoad(key);
+    Assertions.assertThat(storageCredentialCache.getIfPresent(key)).isNotNull();
```
Comment (dimas-b) [inline polaris-core/src/test/java/org/apache/polaris/core/storage/cache/StorageCredentialCacheTest.java:504]: This assert does not prove "Preserves Half Lifetime" (from method name) 🤷

**Review by dimas-b — COMMENTED** [review]
Thanks for this enhancement, @visit2rahul !

File: polaris-core/src/test/java/org/apache/polaris/core/storage/cache/StorageCredentialCacheTest.java [hunk 2]
```diff
@@ -442,6 +443,74 @@ private static List<StorageAccessConfig> getFakeAccessConfigs(int number, boolea
...
+    // The entry must be absent immediately after load.
+    long expiresAt = System.currentTimeMillis() + 120_000;
+    StorageAccessConfig nearExpiryConfig =
+        StorageAccessConfig.builder()
+            .put(StorageAccessProperty.AWS_KEY_ID, "key")
+            .put(StorageAccessProperty.AWS_SECRET_KEY, "secret")
+            .put(
+                StorageAccessProperty.AWS_SESSION_TOKEN_EXPIRES_AT_MS,
+                String.valueOf(expiresAt))
+            .put(StorageAccessProperty.EXPIRATION_TIME, String.valueOf(expiresAt))
```
Comment (dimas-b) [inline polaris-core/src/test/java/org/apache/polaris/core/storage/cache/StorageCredentialCacheTest.java]: This will probably need to be reworked after #5070

File: polaris-core/src/test/java/org/apache/polaris/core/storage/cache/StorageCredentialCacheTest.java (hunk 1 above)
Comment (visit2rahul) [reply polaris-core/src/test/java/org/apache/polaris/core/storage/cache/StorageCredentialCacheTest.java:504]: Good point -- renamed to what the assert actually proves.

File: polaris-core/src/test/java/org/apache/polaris/core/storage/cache/StorageCredentialCacheTest.java (hunk 2 above)
Comment (visit2rahul) [reply polaris-core/src/test/java/org/apache/polaris/core/storage/cache/StorageCredentialCacheTest.java]: done...

**Comment by visit2rahul** [comment]
@dimas-b Morning. Just checking if anything else is needed on this one?

File: polaris-core/src/test/java/org/apache/polaris/core/storage/cache/StorageCredentialCacheTest.java [hunk 3]
```diff
@@ -441,6 +442,68 @@ private static List<StorageAccessConfig> getFakeAccessConfigs(int number, boolea
...
+            realmContext.getRealmIdentifier(),
+            "bufferConfig",
+            true,
+            Set.of("s3://bucket/path"),
+            Set.of(),
+            Optional.empty(),
+            realmConfigWithBuffer,
+            () -> nearExpiryConfig);
+
+    storageCredentialCache.getOrLoad(key);
```
Comment (flyingImer) [inline polaris-core/src/test/java/org/apache/polaris/core/storage/cache/StorageCredentialCacheTest.java:478]: IIUC, this test only verifies that a near-expiry credential is not retained but it does not verify the documented guarantee that clients receive at least the configured remaining validity. 

StorageCredentialCache.getOrLoad directly returns Caffeine LoadingCache.get, and it returns the freshly loaded value even when expireAfterCreate gives it a zero TTL, so the first caller still receives nearExpiryConfig and only later lookups miss. 

Could we ensure getOrLoad never returns an under-buffer credential and assert that behavior here? One option is to validate immediately after key.load(); if the intended contract is only to prevent later cache hits, the configuration description should be narrowed instead.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/cache/StorageCredentialCache.java [hunk 4]
```diff
@@ -86,6 +87,12 @@ private long maxCacheDurationMs(RealmConfig realmConfig) {
     }
   }
 
+  /** Minimum buffer to keep between credential expiry and cache eviction. */
+  private long refreshBufferMs(RealmConfig realmConfig) {
+    return realmConfig.getConfig(FeatureConfiguration.STORAGE_CREDENTIAL_REFRESH_BUFFER_SECONDS)
+        * 1000L;
```
Comment (flyingImer) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/cache/StorageCredentialCache.java:93]: The advertised refresh-buffer range is not enforced. refreshBufferMs accepts any integer: a negative value silently selects the legacy half-life path, while a value at or above the configured credential duration can make every request reload from the provider when credentials are issued at that duration. 

Could we reject values the public description declares invalid and define negative values explicitly, rather than silently changing cache behavior? One option is to accept only 0 <= buffer < STORAGE_CREDENTIAL_DURATION_SECONDS before key.load(), matching the adjacent cache-duration constraint, with negative/equal/greater boundary tests.

**Review by flyingImer — COMMENTED** [review]
I'm aligned with making the cache margin operator-configurable, but I don't think this is merge-ready yet. The positive-buffer contract needs to distinguish cache reuse from what getOrLoad may return, and values outside the supported range need deterministic rejection. Once those behaviors and boundary tests align, no other blocking concern from my side.

File: site/content/in-dev/unreleased/configuration/config-sections/flags-polaris_features.md [hunk 5]
```diff
@@ -540,6 +540,15 @@ The duration of time that vended storage credentials are valid for. Support for
 
 ---
 
+##### `polaris.features."STORAGE_CREDENTIAL_REFRESH_BUFFER_SECONDS"`
+
+Minimum remaining validity (in seconds) that a cached storage credential must have before it is considered stale and evicted. When non-zero, credentials are removed from the cache this many seconds before they expire, ensuring clients always receive credentials with at least this much validity left. When zero (the default), credentials are kept for half of their remaining lifetime, which is the legacy behavior. Must be less than STORAGE_CREDENTIAL_DURATION_SECONDS.
```
Comment (ayushtkn) [inline site/content/in-dev/unreleased/configuration/config-sections/flags-polaris_features.md:545]: > ensuring clients always receive credentials with at least this much validity lef

But getOrLoad delegates directly to `LoadingCache.get():`

```
public StorageAccessConfig getOrLoad(StorageCredentialCacheKey key) {
  return cache.get(key).toAccessConfig();
}
```
When `effectiveTtl` clamps to `0`, Caffeine still returns the freshly loaded entry on that same `get()` call. 

The new test demonstrates this:
`storageCredentialCache.getOrLoad(key);`  // returns cred with only 120s left
`Assertions.assertThat(storageCredentialCache.getIfPresent(key)).isNull();`  // not cached
So with `buffer=300s` and a credential expiring in `120s`, the first caller still receives a nearly-expired credential — only subsequent cache lookups miss. That contradicts the documented contract.

**Review by ayushtkn — COMMENTED** [review]
Thanx @visit2rahul few suggestions

Missing C`HANGELOG.md` entry. This is a user-facing configuration flag. Maybe worth adding

Test coverage gaps:

* No test for the happy path: buffer=300, credential expires in 3600s → entry cached with ~3300s effective TTL
* No test that a second getOrLoad triggers reload when the first result was evicted due to buffer (similar to testCacheEvict)
* No boundary validation tests (blocked on implementation)

**Comment by visit2rahul** [comment]
Thanks both — good catch, the first load wasn't actually protected. Fixed `getOrLoad` to retry once if the fresh load is still under the buffer, and added the range check you both mentioned. Added tests for the happy path, the retry, and the boundary cases, plus the CHANGELOG entry. Also rebased since the branch was pretty far behind main. Let me know if this isn't what you had in mind.

**Review by flyingImer — COMMENTED** [review]
Thanks for the update. The range check addresses my configuration concern. The retry also covers the case where the second credential is fresh, but I still see one gap before merge: getOrLoad returns the second credential after warning even when it is still below the buffer, while the config description and changelog still promise minimum validity. With a 300s buffer and both loads returning credentials with 120s left, the caller still receives one.

Keeping the retry bounded makes sense. Could we either fail after the retry cannot meet the buffer, or document this as best-effort refresh with that fallback? The repeated-near-expiry test should assert the chosen caller-visible outcome and exactly two loads.
